### ISSUE TITLE
--overwrite option doesn't apply to preview if `--preview` specified

### ISSUE THREAD

**Issue by spencerc99** [RhetTbull/osxphotos#526]
Using `--overwrite` in combination with `--preview` doesn't overwrite existing preview images and instead re-exports them. Would expect it to have the same behavior as with exporting normal images.

Let me know if you need me to provide a smaller reproduction or an image to help!

**Comment by RhetTbull** [comment]
See also, #516

**Comment by RhetTbull** [comment]
@all-contributors add @spencerc99 for bug

**Comment by RhetTbull** [comment]
This is fixed in [v0.42.93](https://github.com/RhetTbull/osxphotos/releases/tag/v0.42.93).  Thanks for the bug report!

**Comment by RhetTbull** [comment]
Yeah, that's another bug.  I've reopened this issue.

**Comment by RhetTbull** [comment]
I think this is fixed in [v0.42.94](https://github.com/RhetTbull/osxphotos/releases/tag/v0.42.94).  

Because a preview image will likely exist even if the original image is missing on disk, the preview code can be reached even if the main export code has been skipped (because original image is missing) but that means a lot of the checks (e.g. for duplicate files) that happen in that part of the code also got skipped which is why these two issues happened.  Let me know if you still see any issues!

**Comment by RhetTbull** [comment]
@spencerc99 what's the full command line you're using that leads to this issue?

**Comment by spencerc99** [comment]
here's my full command:
`osxphotos export --download-missing --album fits\ 🧢 --use-photokit --skip-live --skip-original-if-edited --filename "{created.date}" --update --convert-to-jpeg --post-function scripts/transform_exported_fits_into_db.py::post_function --preview --verbose ./fits-export`

I'll try avoiding duplicate names! Will the state DB will recognize that I've already exported all these photos even if the export name changes?

**Comment by RhetTbull** [comment]
> looks like the DB unfortunately does not recognize that

Yes, if you change the file naming scheme, osxphotos won't be able to correlate previously exported photos to the new names.  The export database uses the exported filename as the key for storing metadata about the photo.  This is what allows `--update` to work.

> strangely one preview missing

That is odd.  It is possible the preview file itself was missing (would be unusual but not impossible) and Photos regenerated it.  Because you're using `--use-photokit`, osxphotos will use the native PhotoKit API to do the export so it's possible PhotoKit requested the photo which was not available on the local machine and had to be downloaded from iCloud which then caused Photos to regenerate the missing preview. 

Also, I noticed you're using a `/` in your filename.  That's fine but you should be aware that, while the `/` shows in Finder, in the Terminal, the `/` character, which is the directory separator, will be represented as a `:`.  This happens automatically but you'll need to know this if you are manipulating files in the command line.  For this reason, I prefer a `-` or `_` character in filenames.

**Comment by RhetTbull** [comment]
Thanks for the command line. I'll see if I can re-create the issue you're seeing.  The export code is the most complicated bit of code in osxphotos and it's ripe for refactoring (#462, #529) -- unfortunately I'm in the midst of another project at the moment so this will have to wait a bit.  I've already sketched out a new design though for the export code which should simplify things and make it easier to track down bugs like this issue.

**Comment by spencerc99** [comment]
> Also, I noticed you're using a / in your filename. That's fine but you should be aware that, while the / shows in Finder, in the Terminal, the / character, which is the directory separator, will be represented as a :. This happens automatically but you'll need to know this if you are manipulating files in the command line. For this reason, I prefer a - or _ character in filenames.

Oh interesting. I actually was using `:` as the separator there in my script but I noticed that it converted to `/` in Finder, so just naively switched it to `/` thinking that it would be interchangeable, but looks like I ran into that conversion.

> That is odd. It is possible the preview file itself was missing (would be unusual but not impossible) and Photos regenerated it. Because you're using --use-photokit, osxphotos will use the native PhotoKit API to do the export so it's possible PhotoKit requested the photo which was not available on the local machine and had to be downloaded from iCloud which then caused Photos to regenerate the missing preview.

hmm I see and is the reason it missed it in the first round of exports then because `--download-missing` only applies to originals?

**Comment by spencerc99** [comment]
> Thanks for the command line. I'll see if I can re-create the issue you're seeing. The export code is the most complicated bit of code in osxphotos and it's ripe for refactoring (#462, #529) -- unfortunately I'm in the midst of another project at the moment so this will have to wait a bit. I've already sketched out a new design though for the export code which should simplify things and make it easier to track down bugs like this issue.

no worries appreciate the quick help! using the original filename in my export file name should work for now and makes the end export cleaner too :)

**Comment by RhetTbull** [comment]
@spencerc99 I finally got some time to take a look at this issue.  I have not been able to reproduce the "preview_preview" issue but I did discover a bug associated with `--preview` and `--download-missing`.  Namely, if the original file is missing and thus the `--download-missing` code runs, the code that exports the preview will not be run and thus previews aren't exported for missing photos.  If the export is run again with --update, chances are good the originals will now be in the library (they were downloaded in the previous run) and now the preview code is reached, exporting the previews.  I'll open a new issue for this.  Still looking into the duplicate preview and the --overwrite issue.
